feat: ignore swapfile for running Nvim processes

justinmk opened the pull request:
# Problem:
The swapfile "E325: ATTENTION" dialog is displayed when editing a file already open in another (running) Nvim. Usually this behavior is annoying and irrelevant:
- "Recover" and the other options ("Open readonly", "Quit", "Abort") are almost never wanted.
- swapfiles are less relevant for "multi-Nvim" since 'autoread' is enabled by default.
  - Even less relevant if user enables 'autowrite'.

# Solution:

<img width="444" alt="image" src="https://github.com/neovim/neovim/assets/1359421/87e750f1-6e52-4adb-ae1c-fbe6f52c3a02">

Define a default SwapExists handler which does the following:
1. If the swapfile is owned by a running Nvim process (matching the current user), automatically chooses "(E)dit anyway" (caveat: this creates a new, extra swapfile, which is mostly harmless and ignored except by `:recover` or `nvim -r`.
2. Shows a 1-line "ignoring swapfile..." message.
3. Users can disable the default SwapExists handler via `autocmd! nvim.swapfile`.

justinmk committed:
feat: ignore swapfile for running Nvim processes

Problem:
The swapfile "E325: ATTENTION" dialog is displayed when editing a file
already open in another (running) Nvim. Usually this behavior is
annoying and irrelevant:
- "Recover" and the other options ("Open readonly", "Quit", "Abort") are
  almost never wanted.
- swapfiles are less relevant for "multi-Nvim" since 'autoread' is
  enabled by default.
  - Even less relevant if user enables 'autowrite'.

Solution:
If an existing swapfile is owned by a running Nvim process,
1. automatically choose "Edit anyway" (caveat: this creates a new, extra
   swapfile, which is mostly harmless and ignored except when `:recover`
   or `nvim -r` is explicitly used.
2. show a 1-line "swapfile found, editing anyway..." message
3. user can opt-in to the swapfile dialog using `:confirm edit` or the
   'confirm' option.

TESTS:
- can override with :confirm/'confirm'
- skips dialog when PID is running
- shows dialog when PID is not running

zeertzjq commented:
With this PR the user is no longer be able to use `SwapExists` to set a different default option in this case.

I usually select "Abort" or "Quit" in this case as I usually don't want to have two Nvim instances editing the same file.

MariaSolOs commented on runtime/doc/editing.txt:
```diff
@@ -169,33 +169,30 @@ If you want to keep the changed buffer without saving it, switch on the
+			If [!] is given, unsaved changes in the current buffer
+			are discarded. Without [!] the command fails if there
+			are unsaved changes, unless 'autowriteall' is set and
+			the file can be written.
```
```suggestion
			the file can be written to.
```
English isn't my first language but "can be written" sounds kind of odd to me.

zeertzjq replied:
"write the file" or "write a file" appears everywhere in docs.

MariaSolOs replied:
Fair enough.

MariaSolOs commented on src/nvim/memline.c:
```diff
@@ -334,7 +331,7 @@ int ml_open(buf_T *buf)
 
   // Always sync block number 0 to disk, so we can check the file name in
-  // the swap file in findswapname(). Don't do this for a help files or
+  // the swapfile in findswapname(). Don't do this for a help files or
```
```suggestion
  // the swapfile in findswapname(). Don't do this for help files or
```

MariaSolOs commented on src/nvim/memline.c:
```diff
@@ -167,17 +167,15 @@ enum {
 // This block is built up of single bytes, to make it portable across
 // different machines. b0_magic_* is used to check the byte order and size of
-// variables, because the rest of the swap file is not portable.
+// variables, because the rest of the swapfile is not portable.
```
Haha do you also have very strong opinions about `swapfile` versus `swap file` @justinmk 😆?

justinmk replied:
:) not really, but trying to be consistent at least internally. "swap file name" reads awkwardly. 

I could be persuaded to revert this though.

MariaSolOs replied:
I don't have strong feelings about it, but I agree that "swapfile" reads better.

justinmk committed:
feat: ignore swapfile for running Nvim processes

(This is an alternative to the parent commit which makes the swapfile
dialog opt-in via ":confirm edit".)

Problem:
The swapfile "E325: ATTENTION" dialog is displayed when editing a file
already open in another (running) Nvim. Usually this behavior is
annoying and irrelevant:
- "Recover" and the other options ("Open readonly", "Quit", "Abort") are
  almost never wanted.
- swapfiles are less relevant for "multi-Nvim" since 'autoread' is
  enabled by default.
  - Even less relevant if user enables 'autowrite'.

Solution:
Define a default SwapExists handler which does the following:
1. If the swapfile is owned by a running Nvim process, automatically
   chooses "(E)dit anyway" (caveat: this creates a new, extra swapfile,
   which is mostly harmless and ignored except by `:recover` or `nvim -r`.
2. Shows a 1-line "ignoring swapfile..." message.
3. Users can disable the default SwapExists handler via `autocmd! nvim_swapfile`.

justinmk commented:
Changed approach, updated description.

> I think it's better to expose this information (whether there is a running process editing the file) to `SwapExists` autocommand, and then implement this as a default autocommand.

Done. Currently the pid is exposed via `v:swappid` but I think I'll change it to `v:event.pid`

zeertzjq commented on runtime/lua/vim/_editor.lua:
```diff
@@ -1152,6 +1152,11 @@ function vim._init_default_autocmds()
+  vim.api.nvim_create_autocmd({ 'SwapExists' }, {
+    pattern = '*',
+    group = vim.api.nvim_create_augroup('nvim_swapfile', {}),
+    command = [[if v:swappid > 0 | let v:swapchoice='e' | echo 'E325: ignoring swapfile from Nvim process '..v:swappid | endif]],
```
```suggestion
    command = [[if v:swappid > 0 | let v:swapchoice='e' | echo 'E325: Ignoring swapfile from Nvim process '..v:swappid | endif]],
```

justinmk commented on src/nvim/message.c:
```diff
@@ -475,7 +475,14 @@ void trunc_string(const char *s, char *buf, int room_in, int buflen)
-// Note: Caller of smsg() must check the resulting string is shorter than IOSIZE!!!
+/// Shows a printf-style message with attributes.
+///
+/// Note: Caller must check the resulting string is shorter than IOSIZE!!!
```
Is this comment still relevant? I don't see any callers checking the size.

justinmk commented:
Removing `v:swappid`, changed `swapinfo().pid` to be zero if the process is not running or was from before a reboot.
